File: python/2023/13/pattern.py
from dataclasses import dataclass
from typing import NamedTuple


class Point(NamedTuple):
    x: int
    y: int


@dataclass
class Pattern:
    rocks: list[Point]

    @classmethod
    def build_from_str(cls, input: str) -> "Pattern":
        rocks = []
        for y, line in enumerate(input.splitlines()):
            for x, char in enumerate(line):
                if char == "#":
                    rocks.append(Point(x, y))
        return cls(rocks)
    

    def find_horizontal_reflection(self) -> int | None:
        # Where is the line over which this pattern is symmetrical?
        max_x = max(rock.x for rock in self.rocks)
        max_y = max(rock.y for rock in self.rocks)
        for x_line in range(0, max_x):
            # The line actually runs between two points.
            x_line = x_line + 0.5
            # Start at this point and work to the right.
            x = x_line + 0.5
            reflected_x = x_line - 0.5
            all_match = True
            while x <= max_x and reflected_x >= 0:
                for y in range(0, max_y + 1):
                    is_rock = Point(x, y) in self.rocks
                    is_reflected_rock = Point(reflected_x, y) in self.rocks
                    if is_rock != is_reflected_rock:
                        all_match = False
                        break
                x += 1
                reflected_x -= 1
            if all_match:
                return int(x_line - 0.5)
        return None
    
    def find_vertical_reflection(self) -> int | None:
        # Where is the line over which this pattern is symmetrical?
        max_x = max(rock.x for rock in self.rocks)
        max_y = max(rock.y for rock in self.rocks)
        for y_line in range(0, max_y):
            # The line actually runs between two points.
            y_line = y_line + 0.5
            # Start at this point and work down.
            y = y_line + 0.5
            reflected_y = y_line - 0.5
            all_match = True
            while y <= max_y and reflected_y >= 0:
                for x in range(0, max_x + 1):
                    is_rock = Point(x, y) in self.rocks
                    is_reflected_rock = Point(x, reflected_y) in self.rocks
                    if is_rock != is_reflected_rock:
                        all_match = False
                        break
                y += 1
                reflected_y -= 1
            if all_match:
                return int(y_line - 0.5)
        return None
    
    def find_reflect_score(self) -> int:
        vert_score = self.find_vertical_reflection()
        if vert_score is not None:
            return (vert_score + 1) * 100
        horiz_score = self.find_horizontal_reflection()
        if horiz_score is not None:
            return horiz_score + 1
        raise RuntimeError("no symmetry")
    
    def __str__(self) -> str:
        max_x = max(rock.x for rock in self.rocks)
        max_y = max(rock.y for rock in self.rocks)
        lines = []
        for y in range(0, max_y + 1):
            line = []
            for x in range(0, max_x + 1):
                if Point(x, y) in self.rocks:
                    line.append("#")
                else:
                    line.append(".")
            lines.append("".join(line))
        return "\n".join(lines)

File: rust/src/y2023/d13/pattern.rs
use std::collections::HashSet;

#[derive(PartialEq, Eq, Hash)]
pub struct Point {
    pub x: usize,
    pub y: usize,
}

pub struct Pattern {
    pub points: HashSet<Point>,
}

impl Pattern {
    pub fn new_from_str(s: &str) -> Self {
        let mut points = HashSet::new();
        for (y, line) in s.lines().enumerate() {
            for (x, c) in line.chars().enumerate() {
                if c == '#' {
                    points.insert(Point { x, y });
                }
            }
        }
        Self { points }
    }

    pub fn to_str(&self) -> String {
        let mut s = String::new();
        let max_x = self.points.iter().map(|p| p.x).max().unwrap();
        let max_y = self.points.iter().map(|p| p.y).max().unwrap();
        for y in 0..=max_y {
            for x in 0..=max_x {
                if self.points.contains(&Point { x, y }) {
                    s.push('#');
                } else {
                    s.push('.');
                }
            }
            s.push('\n');
        }
        return s;
    }

    pub fn find_horizontal_symmetry(&self, mismatches: usize) -> Option<usize> {
        // Where is the line over which this pattern is symmetrical?
        let max_x = self.points.iter().map(|p| p.x).max().unwrap();
        let max_y = self.points.iter().map(|p| p.y).max().unwrap();
        for x_line in 0..max_x {
            // The line actually runs *between* two points.
            let x_line = x_line as f64 + 0.5;
            // Start at this point and work to the right.
            let mut x = (x_line + 0.5) as usize;
            let mut reflected_x = (x_line - 0.5) as i32;
            let mut mismatches_found = 0;
            while x <= max_x && reflected_x >= 0 {
                for y in 0..=max_y {
                    let is_rock = self.points.contains(&Point { x, y });
                    let is_reflected_rock = self.points.contains(&Point {
                        x: reflected_x as usize,
                        y,
                    });
                    if is_rock != is_reflected_rock {
                        mismatches_found += 1;
                        if mismatches_found > mismatches {
                            // Once we've found too many mismatches, there's no point finding more.
                            break;
                        }
                    }
                }
                x += 1;
                reflected_x -= 1;
            }
            if mismatches_found == mismatches {
                return Some((x_line - 0.5) as usize);
            }
        }
        None
    }

    pub fn find_vertical_symmetry(&self, mismatches: usize) -> Option<usize> {
        // Where is the line over which this pattern is symmetrical?
        let max_x = self.points.iter().map(|p| p.x).max().unwrap();
        let max_y = self.points.iter().map(|p| p.y).max().unwrap();
        for y_line in 0..max_y {
            // The line actually runs *between* two points.
            let y_line = y_line as f64 + 0.5;
            // Start at this point and work to the right.
            let mut y = (y_line + 0.5) as usize;
            let mut reflected_y = (y_line - 0.5) as i32;
            let mut mismatches_found = 0;
            while y <= max_y && reflected_y >= 0 {
                for x in 0..=max_x {
                    let is_rock = self.points.contains(&Point { x, y });
                    let is_reflected_rock = self.points.contains(&Point {
                        x,
                        y: reflected_y as usize,
                    });
                    if is_rock != is_reflected_rock {
                        mismatches_found += 1;
                        if mismatches_found > mismatches {
                            // Once we've found too many mismatches, there's no point finding more.
                            break;
                        }
                    }
                }
                y += 1;
                reflected_y -= 1;
            }
            if mismatches_found == mismatches {
                return Some((y_line - 0.5) as usize);
            }
        }
        None
    }

    pub fn find_reflect_score(&self, mismatches: usize) -> Result<usize, Box<dyn std::error::Error>> {
        let vert_score = self.find_vertical_symmetry(mismatches);
        if let Some(vert_score) = vert_score {
            return Ok((vert_score + 1) * 100);
        }
        let horiz_score = self.find_horizontal_symmetry(mismatches);
        if let Some(horiz_score) = horiz_score {
            return Ok(horiz_score + 1);
        }
        Err("no symmetry".into())
    }
}